cpp/src/binaryop/compiled/util.cpp

```diff
@@ -0,0 +1,133 @@
+        case binary_operator::FLOOR_DIV:            return call<ops::FloorDiv, TypeOut>();
+        case binary_operator::MOD:                  return call<ops::Mod, TypeOut>();
+        case binary_operator::PYMOD:                return call<ops::PyMod, TypeOut>();
+        case binary_operator::POW:                  return call<ops::Pow, TypeOut>();
+        case binary_operator::BITWISE_AND:          return call<ops::BitwiseAnd, TypeOut>();
+        case binary_operator::BITWISE_OR:           return call<ops::BitwiseOr, TypeOut>();
+        case binary_operator::BITWISE_XOR:          return call<ops::BitwiseXor, TypeOut>();
+        case binary_operator::SHIFT_LEFT:           return call<ops::ShiftLeft, TypeOut>();
+        case binary_operator::SHIFT_RIGHT:          return call<ops::ShiftRight, TypeOut>();
+        case binary_operator::SHIFT_RIGHT_UNSIGNED: return call<ops::ShiftRightUnsigned, TypeOut>();
+        case binary_operator::LOG_BASE:             return call<ops::LogBase, TypeOut>();
+        case binary_operator::ATAN2:                return call<ops::ATan2, TypeOut>();
+        case binary_operator::PMOD:                 return call<ops::PMod, TypeOut>();
+        /*
+        case binary_operator::NULL_MAX:             return call<ops::NullMax, TypeOut>();
+        case binary_operator::NULL_MIN:             return call<ops::NullMin, TypeOut>();
+        case binary_operator::GENERIC_BINARY:       return call<ops::UserDefinedOp, TypeOut>();
+        */
+        default:                                    return false;
+          // clang-format on
+      }
+    }
+  };
+
+  template <typename BinaryOperator, typename TypeLhs, typename TypeRhs>
+  inline constexpr bool bool_op(data_type out)
+  {
+    return out.id() == type_id::BOOL8 and
+           is_binary_operation_supported<BinaryOperator>{}.template operator()<TypeLhs, TypeRhs>();
+  }
+  template <typename TypeLhs, typename TypeRhs>
+  inline constexpr bool operator()(data_type out, binary_operator op)
+  {
+    switch (op) {
+      // output type should be bool type.
+      case binary_operator::LOGICAL_AND: return bool_op<ops::LogicalAnd, TypeLhs, TypeRhs>(out);
+      case binary_operator::LOGICAL_OR: return bool_op<ops::LogicalOr, TypeLhs, TypeRhs>(out);
+      case binary_operator::EQUAL: return bool_op<ops::Equal, TypeLhs, TypeRhs>(out);
+      case binary_operator::NOT_EQUAL: return bool_op<ops::NotEqual, TypeLhs, TypeRhs>(out);
+      case binary_operator::LESS: return bool_op<ops::Less, TypeLhs, TypeRhs>(out);
+      case binary_operator::GREATER: return bool_op<ops::Greater, TypeLhs, TypeRhs>(out);
+      case binary_operator::LESS_EQUAL: return bool_op<ops::LessEqual, TypeLhs, TypeRhs>(out);
+      case binary_operator::GREATER_EQUAL: return bool_op<ops::GreaterEqual, TypeLhs, TypeRhs>(out);
+      case binary_operator::NULL_EQUALS: return bool_op<ops::NullEquals, TypeLhs, TypeRhs>(out);
```

Comment:
Why are these operators separate from the other ones?

Reply:
These operators results are always bool. and output_dtype is restricted to bool.
These are separated from other ones to reduce compile time. (only 2 type dispatchers required)

Reply:
For util.cpp, Parallel compile time is different from single file compile time. It saves ~25 seconds.